### PR TITLE
Show the Headings that are declarable using commodity View

### PR DESCRIPTION
**Pull request by alexdesi**
### What?
This PR shows the Headings that are declarable using commodity View

For example: URL /headings/0903 -> URL commodities/0903000000

### Jira link
https://transformuk.atlassian.net/browse/HOTT-1129

### Why?
To improve User usability and navigation of the commodities and headings.

### CODE REVIEW

**Comment by alexdesi** [comment]
When heading is declarable: 
(note: heading is not present in the breadcrumb)
Screenshot 2021-11-26 at 09 20 12
---
When heading is NOT declarable:
Screenshot 2021-11-26 at 09 20 40
:

File: app/controllers/commodities_controller.rb [hunk 1]
```diff
@@ -26,6 +25,17 @@ def commodities_by_code
...
+      heading_id = params[:id].slice(0, 4)
+      heading = HeadingPresenter.new(Heading.find(heading_id, query_params))
+
+      if heading.declarable? # stay in commodity page
+        @commodities[:uk] = heading
+
+        return
+      end
+    end
+
```
Comment (alexdesi) [inline app/controllers/commodities_controller.rb]: This is like a guard_clause, I am thinking to move this to a different `before_action`

File: app/controllers/commodities_controller.rb [hunk 2]
```diff
@@ -26,6 +25,17 @@ def commodities_by_code
   def fetch_commodities
     @commodities ||= {}
 
+    if is_heading_id?
+      heading_id = params[:id].slice(0, 4)
```
Comment (willfish) [inline app/controllers/commodities_controller.rb]: The last 6 digits should be 0 for a heading :+1:

File: app/controllers/goods_nomenclatures_controller.rb [hunk 3]
```diff
@@ -1,6 +1,10 @@
 class GoodsNomenclaturesController < ApplicationController
   rescue_from Faraday::ResourceNotFound, with: :find_relevant_goods_code_or_fallback
 
+  def is_heading_id?
```
Comment (jebw) [inline app/controllers/goods_nomenclatures_controller.rb]: This feels like business logic? Should this be a class method on either the `Heading` or the `GoodsNomenclature models`

Comment (jebw) [reply app/controllers/goods_nomenclatures_controller.rb]: If not, it should probably be in the `private` section

Comment (jebw) [reply app/controllers/goods_nomenclatures_controller.rb]: Just looked further - this is a duplication of the logic in `GoodsNomenclature#heading?` - I think this should definitely become a class method at `GoodsNomenclature.is_heading_id?(id)`, and `GoodsNomenclature#heading?` needs changing to delegate to logic in the class method

**Review by jebw — CHANGES_REQUESTED** [review]
Looks great - only query is the duplicating business logic out the `GoodsNomenclature` model - instead that should instead be refactored/reused

File: app/controllers/goods_nomenclatures_controller.rb (hunk 3 above)
Comment (alexdesi) [reply app/controllers/goods_nomenclatures_controller.rb]: I agree, I'll dry the logic.

File: app/controllers/commodities_controller.rb [hunk 4]
```diff
@@ -23,17 +22,43 @@ def commodities_by_code
     Commodity.by_code(search_term).sort_by(&:code)
   end
 
+  def fetch_headings_or_commodities
+    if heading_from_param&.declarable?
+      fetch_headings
+    else
+      fetch_commodities
+    end
+  end
```
Comment (alexdesi) [inline app/controllers/commodities_controller.rb:30]: I added this to avoid adding another conditional before_action filter, and try to be more explicit.

**Comment by alexdesi** [comment]
The `fetch_commodities` method was getting too complex,
so I added a similar method `fetch_headings` to support it (and I simplified it too).

**Review by jebw — APPROVED** [review]
LGTM :+1:

File: app/helpers/declarable_helper.rb [hunk 5]
```diff
@@ -1,6 +1,14 @@
 module DeclarableHelper
   def classification_description(commodity)
```
Comment (willfish) [inline app/helpers/declarable_helper.rb]:
```suggestion
  def classification_description(goods_nomenclature)
```

File: app/helpers/declarable_helper.rb [hunk 6]
```diff
@@ -1,6 +1,14 @@
...
-    descriptions = [commodity.heading, commodity.ancestors, commodity]
+    tree_items = if commodity.heading?
```
Comment (willfish) [inline app/helpers/declarable_helper.rb]:
```suggestion
    tree_items = if goods_nomenclature.heading?
```

File: app/helpers/declarable_helper.rb [hunk 7]
```diff
@@ -1,6 +1,14 @@
...
+                   [commodity]
```
Comment (willfish) [inline app/helpers/declarable_helper.rb]:
```suggestion
                   [goods_nomenclature]
```

File: app/helpers/declarable_helper.rb [hunk 8]
```diff
@@ -1,6 +1,14 @@
...
+                 else
+                   [commodity.heading,
```
Comment (willfish) [inline app/helpers/declarable_helper.rb]:
```suggestion
                   [goods_nomenclature.heading,
```

File: app/helpers/declarable_helper.rb [hunk 9]
```diff
@@ -1,6 +1,14 @@
...
+                    commodity.ancestors,
```
Comment (willfish) [inline app/helpers/declarable_helper.rb]:
```suggestion
                    goods_nomenclature.ancestors,
```

File: app/helpers/declarable_helper.rb [hunk 10]
```diff
@@ -1,6 +1,14 @@
...
+                    commodity]
```
Comment (willfish) [inline app/helpers/declarable_helper.rb]:
```suggestion
                    goods_nomenclature]
```

File: app/helpers/declarable_helper.rb (hunk 6 above)
Comment (willfish) [reply app/helpers/declarable_helper.rb]: Or declarable

File: app/controllers/commodities_controller.rb [hunk 11]
```diff
@@ -21,6 +21,33 @@ def commodities_by_code
...
+      @commodities[:uk] = heading_from_param
+      @commodities[:xi] = nil
+    else
+      @commodities[:xi] = heading_from_param
+      @commodities[:uk] = TradeTariffFrontend::ServiceChooser.with_source(:uk) { heading_from_param }
+    end
+  end
+
+  def heading_from_param
+    @heading_from_param ||= if GoodsNomenclature.is_heading_id?(params[:id])
```
Comment (willfish) [inline app/controllers/commodities_controller.rb:45]: Separately, we should update the places that ask this question for consistency

File: app/helpers/declarable_helper.rb (hunk 6 above)
Comment (willfish) [reply app/helpers/declarable_helper.rb]: So most of what's being asked here is a bit greedy on the goods_nomenclature model. I wonder if tree_items should be pushed down to the individual model?

Comment (willfish) [reply app/helpers/declarable_helper.rb]: e.g. Heading#tree_items Commodity#tree_items

Comment (willfish) [reply app/helpers/declarable_helper.rb]: Then you don't have to know about whether the declarable is a heading

File: app/models/goods_nomenclature.rb [hunk 12]
```diff
@@ -65,4 +65,8 @@ def rules_of_origin(*args, **kwargs)
   def rules_of_origin_rules(...)
     rules_of_origin(...).flat_map(&:rules)
   end
+
+  def self.is_heading_id?(goods_code_id)
```
Comment (willfish) [inline app/models/goods_nomenclature.rb]:
```suggestion
  def self.is_heading_id?(goods_nomenclature_item_id)
```

File: app/views/commodities/_header.html.erb [hunk 13]
```diff
@@ -1,4 +1,8 @@
 <header>
+  <span class="govuk-caption-m">
+    <%= trade_tariff_heading %>
```
Comment (willfish) [inline app/views/commodities/_header.html.erb:3]: I guess this was missing :+1:

**Review by willfish — APPROVED** [review]

**Pull request merged**